Common/cpp/SharedItems/Shareables.cpp

```diff
@@ -0,0 +1,85 @@
+#include "Shareables.h"
+
+using namespace facebook;
+
+namespace reanimated {
+
+CoreFunction::CoreFunction(JSRuntimeHelper *_runtimeHelper, const jsi::Value &workletValue) : runtimeHelper(_runtimeHelper) {
+  jsi::Runtime &rt = *_runtimeHelper->rnRuntime;
+  auto workletObject = workletValue.asObject(rt);
+  rnFunction = std::make_shared<jsi::Function>(workletObject.asFunction(rt));
+  functionBody = workletObject.getProperty(rt, "asString").asString(rt).utf8(rt);
+  location = workletObject.getProperty(rt, "__location").asString(rt).utf8(rt);
+}
+
+jsi::Value CoreFunction::call(jsi::Runtime &rt, const jsi::Value &arg0) {
+  if (runtimeHelper->isUIRuntime(rt)) {
+    if (uiFunction == nullptr) {
+      // maybe need to initialize UI Function
+      auto codeBuffer =
+          std::make_shared<const jsi::StringBuffer>("(" + functionBody + ")");
+      uiFunction = std::make_shared<jsi::Function>(
+          rt.evaluateJavaScript(codeBuffer, location)
+              .asObject(rt)
+              .asFunction(rt));
+    }
+    return uiFunction->call(rt, arg0);
+  } else {
+    // running on the main RN runtime
+    return rnFunction->call(rt, arg0);
+  }
+}
+
+std::shared_ptr<Shareable> extractShareableOrThrow(
+    jsi::Runtime &rt,
+    const jsi::Value &maybeShareableValue) {
+  if (maybeShareableValue.isObject()) {
+    auto object = maybeShareableValue.asObject(rt);
+    if (object.isHostObject<ShareableJSRef>(rt)) {
+      return object.getHostObject<ShareableJSRef>(rt)->value;
+    }
+  }
+  throw std::runtime_error("expecing the object to be of type ShareableJSRef");
+}
+
+Shareable::~Shareable() {
+}
+
+jsi::Value RetainingShareable::getJSValue(jsi::Runtime &rt) {
+  jsi::Value value;
+  if (&rt == hostRuntime) {
+    // TODO: it is suboptimal to generate new object every time getJS is called on host runtime – the
+    // object we are generating already exists and we should possibly just grab a hold of that object
+    // and use it here instead of creating a new JS representation.
+    // As far as I understand the only caase where this can be realistically called this way is when
+    // a shared value is created and then accessed on the same runtime
+    return toJSValue(rt);
+  } else if (remoteValue == nullptr || (value = remoteValue->lock(rt)).isUndefined()) {
```

Comment:
What's going on here with `(value = remoteValue->lock(rt))` inside the conditional? I guess that it affects the return in line 61, is this intended? This may be easily overlooked, maybe let's add another if/else inside this branch?